### ISSUE TITLE
MapNamingStrategy vs. ArrayMapNamingStrategy

### ISSUE THREAD

**Issue by adamlundrigan** [zendframework/zend-hydrator#9]
Do we need both?  From what I can deduce (from source and [the PR that introduced ArrayMapNamingStrategy](https://github.com/zendframework/zf2/pull/6197)) they're nearly identical except:
- MapNamingStrategy's argument is a hydration map and ArrayMapNamingStrategy is an extraction map.  Both then flip the array to get the mapping for the opposite direction. 
- MapNamingStrategy allows overriding the reverse map, in case you want an asymmetrical mapping.

**Comment by froschdesign** [comment]
@adamlundrigan 
> Do we need both?

I see no reason for this. Because:

>  …they're nearly identical except

Right.

**Comment by weierophinney** [comment]
What's interesting is that they operate in the exact opposite fashion with regards to the required initial constructor argument.

With `ArrayMapNamingStrategy`, that initial argument is used for extraction, and flipped for hydration.

With `MapNamingStrategy`, that initial argument is used for hydration, and flipped for extraction (unless a second argument is provided; that's used for extraction in that case).

As such, we would need to do a constructor that looks like this:

```php
function (?array $extractionMap = null, ?array $hydrationMap = null)
```

I'll start work on this immediately.

**Comment by weierophinney** [comment]
Fixed with #82, and will be shipped with 3.0.0.
